### ISSUE TITLE
epoll_wait() gets stuck on Windows 10

### ISSUE THREAD

**Issue by linusyang** [piscisaureus/wepoll#5]
I tried a minimal example (an [echo server](https://gist.github.com/linusyang/98c46cb4015f346594b4bf4b1b5a39fc)) using wepoll library (v1.3.0). It works well on Windows 7 SP1 (32-bit). But `epoll_wait()` blocks the program on Windows 10 (1709, 16299.125, 64-bit).  I tried to debug using GDB, it seems `GetQueuedCompletionStatusEx` gets stuck in `_ep_port_poll` and never returns (even if I added a finite timeout instead of `-1`). The compiler is MinGW w64 GCC v7.3.0 (`x86_64-w64-mingw32-gcc echo.c wepoll.c -o echo -g -lws2_32`). Any idea on the problem? Thanks!

**Comment by piscisaureus** [comment]
@linusyang 

I tried to run your example on windows 10, and it works for me. That is, when I connect to it using `nc localhost 10007`, it will echo back what I send to it.

epoll_wait() should always return if you pass a small positive value for time-out. However you say you pass, `-1`, which means epoll_wait() should block forever or until there are any events.

Are you sure you are actually able to connect to your echo server? How are you connecting to it?

**Comment by piscisaureus** [comment]
@linusyang 

Also note that you are not checking the return value from `send()`. Since the socket is nonblocking, send() could be failing with an WSAEWOULDBLOCK error code, which could mean not all the data you expect is echoed back.

**Comment by linusyang** [comment]
Thanks for your reply!

> Are you sure you are actually able to connect to your echo server? How are you connecting to it?

I was using `telnet` to connect to the server. I also tried `nc` in msys2 and WSL and it does not work either. The same code runs well in WSL and I can use `nc` to interact with the echo server. I [changed the example a bit](https://gist.github.com/linusyang/98c46cb4015f346594b4bf4b1b5a39fc/revisions) to show the incoming message.

> However you say you pass, -1, which means epoll_wait() should block forever or until there are any events.

I understand `-1` means blocking indefinitely, which is an expected behaviour. That is how `epoll_wait` works. My point is that `epoll_wait` should be unblocked when there is an incoming connection. But in my tests, it never returns even if I pinged the server. I was wrong about the finite timeout -- the function actually returns, but it still cannot accept any connection during the timeout (say 5000 ms), which just returns 0 and exits.

> Also note that you are not checking the return value from send().

The example I use was quite simple and directly obtained from the web. I just want to test `epoll` functions in wepoll library. The example works well on Linux VMs, and even WSL. And in my tests, the issue is caused by `epoll_wait` that cannot receive incoming connections, but not `send`.

Let me provide the compiled binary for reference:
https://drive.google.com/file/d/1uo6KawVZPbGZ-Fi0TAHFJKx69rzxI1KS/view?usp=sharing
which was compiled using the following command:

    x86_64-w64-mingw32-gcc echo.c wepoll.c -o echo -g -lws2_32 -static

**Comment by piscisaureus** [comment]
@linusyang I tried the binary you sent me, and it works for me: https://snag.gy/QRuc3J.jpg
(I had to dismiss a firewall popup though, maybe it's worthwhile double checking that).

Can you share the output of the following two commands? `netstat -ab`, and `netsh winsock show catalog`. Maybe that'll shed some light on what is going on.

**Comment by linusyang** [comment]
@piscisaureus Thanks a lot for your help! Your hint solves my problem. I finally figure out that there are a lot of (possibly bad) LSPs in the WinSock catalog. I removed some of them (e.g., Antivirus, Proxifier) and the echo.exe now works like a charm. Now I will try to prevent those malicious LSP dlls from hooking into my application.

**Comment by linusyang** [comment]
@piscisaureus Thanks to your comment at libuv/libuv#972, I have finally figured out the solution to bypassing the LSP hijacking issues. The workaround is to always choose the default MSAFD providers by calling WSASocket. I have [updated](https://gist.github.com/linusyang/98c46cb4015f346594b4bf4b1b5a39fc/revisions#diff-1cbc2af4fc8b540e4728e99e99e74928) the echo server example, which uses `WSCEnumProtocols` to list current providers and choose the default MSAFD ones by the hardcoded GUIDs from wepoll.c. Now the example works perfectly even with all nasty LSP dlls installed.

Thanks a lot for helping me debug this issue!

**Comment by piscisaureus** [comment]
@linusyang 
Ok, happy you figured it out.
I'm a little surprised though, wepoll tries really hard to sidestep LSPs, but clearly it isn't succeeding. At least epoll_ctl() should return -1 to indicate that the socket isn't supported. Can you do me a favor and check that that's working as it should?

**Comment by piscisaureus** [comment]
@linusyang Can you try this patch, also: https://github.com/piscisaureus/wepoll/commit/397977470451761e3297b765799a3ce3b129bf20

After applying it wepoll should always work, even if you have LSPs installed and not bypassed.

**Comment by linusyang92** [comment]
@piscisaureus 
> At least epoll_ctl() should return -1 to indicate that the socket isn't supported. Can you do me a favor and check that that's working as it should?

You're correct. If I use the old `socket()` way (which loads non-default LSP dlls), `epoll_ctl()` returns -1.

> Can you try this patch, also: 4e149da
> After applying it wepoll should always work, even if you have LSPs installed and not bypassed.

The patch works like a charm! The client code in echo server doesn't need to change a single line. I can use the old `socket()` method and `epoll_ctl()` returns 0. That's perfect.

**Comment by linusyang** [comment]
@piscisaureus

> At least epoll_ctl() should return -1 to indicate that the socket isn't supported. Can you do me a favor and check that that's working as it should?

You're correct. If I use the old `socket()` way (which loads non-default LSP dlls), `epoll_ctl()` returns -1.

> Can you try this patch, also: 4e149da
> After applying it wepoll should always work, even if you have LSPs installed and not bypassed.

The patch works like a charm! The client code in echo server doesn't need to change at all. I can use the old `socket()` method and `epoll_ctl()` returns 0. That's perfect.

Edit: change account to reply this message.

**Comment by piscisaureus** [comment]
@linusyang Thanks for testing the patch. I'm not going to land it as-is - I think it's not very efficient to  query the winsock catalog every time a socket is added to an epoll set. I'll land a proper patch in a few days.

**Comment by linusyang** [comment]
@piscisaureus I have successfully used the wepoll library as a backend of libev on Windows, which makes it easy to port shadowsocks-libev to Windows (previously it uses the `select()` backend of libev). I am using your LSP protocol chain patch at the moment and will update to formal patches later. Thanks again for the awesome project!

**Comment by piscisaureus** [comment]
@linusyang 

> I have successfully used the wepoll library as a backend of libev on Windows, which makes it easy to port shadowsocks-libev to Windows (previously it uses the select() backend of libev). 

Great to hear you were able to put it to good use.
